text/287-cli-deprecation-warnings.md

```diff
@@ -0,0 +1,224 @@
+unless a test is explicitly checking a deprecated API as a regression test,
+in which case we should add an API for turning that option off for those tests.
+
+### Is this a breaking change?
+
+No.
+
+### What are the drawbacks of this solution?
+
+1. Requires changes to JSII.
+2. Does not allow distinguishing between the customer using deprecated APIs,
+  and the library code using deprecated APIs.
+  We can alleviate this problem by turning on the `CDK_DEPRECATED_ERROR` environment variable by default in our tests,
+  which will be a forcing function for us to stop using deprecated APIs in the Construct Library.
+3. We won't be able to register warnings for accessing deprecated static constants
+  (this includes enums).
+
+### What alternative solutions did you consider?
+
+There are many alternatives that were considered,
+but were, for various reasons, discarded:
+
+#### 1. Manual code for emitting the warnings
+
+Instead of modifying JSII,
+we could simply write the code inside the TypeScript APIs "manually",
+emitting warnings for any deprecated elements being used.
+
+This has the advantage of being the simplest solution,
+but has been discarded because of the large effort,
+and the fact that there's no way for us to verify that code has been added
+(and added correctly).
+
+#### 2. TypeScript decorators
+
+We can use [TypeScript decorators](https://www.typescriptlang.org/docs/handbook/decorators.html)
+to enhance the runtime JavaScript code with our custom logic.
+We can add an `awslint` rule that enforces that every element with the `@deprecated`
+JSDoc tag also needs to have the `@deprecated` decorator present.
+While decorators are still considered an experimental feature,
+our init templates and JSII enable them in `tsconfig.json`,
+which means the vast majority of our customers will have them enabled as well.
+
+The main advantage of this solution is that changes are needed only in the CDK project.
+
+Disadvantages of this solution:
+
+1. It won't be possible to reliably get the module the deprecated element belongs to (only the name of the element) --
+  the decorator is simply a JavaScript function defined in `@aws-cdk/core`,
+  and doesn't have (easy) access to the information on what module a given object it's called on belongs to.
+2. TypeScript does not allow decorators on interfaces (neither on the interface directly, nor on any of its properties).
+  This means we will not be able to register warnings for structs, or struct properties.
+3. We won't be able to register warnings for accessing deprecated static constants
+   (this includes enums).
+
+#### 3. Rely on language-specific deprecation warning mechanisms
+
+JSII emits deprecated elements with the correct language-specific mechanisms in place
+(for example, the `@Deprecated` annotation for Java).
+We could simply piggyback on those,
+instead of writing our own.
+
+Advantages of this solution:
+
+1. Minimal development effort on our side.
```

Comment:
Advantage missed: static analysis instead of runtime analysis allows listing of all deprecated API usages in the code base, **not the just ones that HAPPEN to be executed during a particular run**

Reply:
Added.